tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,269 @@
+// vector of the slice sources. Otherwise returns nullopt.
+absl::optional<std::vector<HloInstruction*>>
+FindSourceOperandsOfSlicesForElementwiseOperation(const HloInstruction* inst) {
+  CHECK(inst->IsElementwise());
+
+  // Check that all operands are slices.
+  if (absl::c_any_of(inst->operands(),
+      [](const HloInstruction* operand) {
+        return operand->opcode() != HloOpcode::kSlice;
+      })) {
+    return absl::nullopt;
+  }
+
+  // Check that all slices are from the same indices of slice sources with
+  // compatible shapes.
+  const HloInstruction* slice0 = inst->operand(0);
+  if (absl::c_any_of(absl::MakeSpan(inst->operands()).subspan(1),
+      [slice0](const HloInstruction* slice) {
+        return !SameSliceConfiguration(slice0, slice)
+            || !ShapeUtil::CompatibleIgnoringElementType(
+                slice0->operand(0)->shape(), slice->operand(0)->shape());
+      })) {
+    return absl::nullopt;
+  }
+
+  // Construct and return a vector of the slice sources.
+  std::vector<HloInstruction*> slice_sources;
+  absl::c_transform(inst->operands(), std::back_inserter(slice_sources),
+      [](HloInstruction* slice) { return slice->mutable_operand(0); });
+  return slice_sources;
+}
+
+// Given an instruction with a slice of slice_sources[i] as its ith operand,
+// returns true if peer_candidate is a peer of the instruction that meets the
+// following requirements:
+// 1) Peer_candidate has the same opcode as the given instruction.
+// 2) The ith operand of peer_candidate is a slice of slice_sources[i].
+// 3) All operands of peer_candidate are slices taken from the same indices.
+bool IsPeerOperation(const HloInstruction* inst,
+    const HloInstruction* peer_candidate,
+    absl::Span<HloInstruction* const> slice_sources) {
+  // Check opcode, user count and operand count.
+  if (peer_candidate->opcode() != inst->opcode() ||
+      peer_candidate->user_count() == 0 ||
+      peer_candidate->operand_count() != inst->operand_count())
+    return false;
+
+  // Checks slices:
+  const HloInstruction* operand_slice0 = peer_candidate->operand(0);
+  for (int64 j = 0; j < slice_sources.size(); ++j) {
+    // The operands of each operation are slices taken from the same indices of
+    // bigger tensors with the same dimensions. The corresponding operands of
+    // all operations are slices taken from the same bigger tensors.
+    const HloInstruction* operand_slice = peer_candidate->operand(j);
+    if (operand_slice->opcode() != HloOpcode::kSlice ||
+        operand_slice->operand(0) != slice_sources[j] ||
+        !SameSliceConfiguration(operand_slice0, operand_slice)) {
+      return false;
+      break;
+    }
+  }
+  return true;
+}
+
+// Compares the cost of implementing one elementwise operations on the
+// slice_sources with the cost of implementing all the individual elementwise
+// operations in operations_on_slices and returns true if the former is less
+// expensive.
```

Comment:
We had some offline conversation about the following transformation. Can we add this to the description for the routine.

Currently we don't support the following transformation because we don't have such a use case yet.
Transform
p = f32[20] parameter(0)
 a = f32[8] slice(p), slice=[0:8]
aa = add(a, a)
b = f32[7] slice(p), slice=[2:9]
bb = add(b, b)

to
p = f32[20] parameter(0)
x = f32[9] slice(p), slice=[0:8]
xx = add(x,x)
 aa = f32[8] slice(xx), slice=[0:8]
bb = f32[7] slice(xx), slice=[2:9]

Reply:
done